pkg/secrets/aws_secrets_manager.go

```diff
@@ -0,0 +1,137 @@
+}
+
+func (a *AwsSecretsManagerDecrypter) Decrypt() (string, error) {
+	secretValue, err := a.awsSecretsManagerClient.FetchSecret(a.region, a.secretName)
+	if err != nil {
+		return "", err
+	}
+
+	if a.isFile { // The secret is assumed to be a file so extract the binary data from the secretValue
+		if len(secretValue.SecretBinary) > 0 { // if the binary data has bytes then its a binary blob
+			return parseBinaryFile(secretValue)
+		} else { // else assume its a plaintext payload for a new file.
+			return parsePlaintextFile(secretValue)
+		}
+	} else if a.secretKey != "" { // The secret is assumed to be a k,v pair return the v
+		return parseSecretKVPair(secretValue, &a.secretKey)
+	} else { // The secret is assumed to be a plaintext value return the value
+		return parseSecretValue(secretValue)
+	}
+}
+
+func (a *AwsSecretsManagerDecrypter) IsFile() bool {
+	return a.isFile
+}
+
+func (a *AwsSecretsManagerDecrypter) parse(params string) error {
+	// Parse the user supplied params to get a map of the k,v pairs
+	var data = make(map[string]string)
+	entries := strings.Split(params, "!")
+	if len(entries) < 2 {
+		return fmt.Errorf(GenericMalformedKeyError)
+	}
+	for _, entry := range entries {
+		kvPair := strings.Split(entry, ":")
+		if len(kvPair) != 2 {
+			return fmt.Errorf(GenericMalformedKeyError)
+		}
+		data[kvPair[0]] = kvPair[1]
+	}
+
+	// Validate and save the required / known fields
+	if region := data[Region]; region != "" {
+		a.region = region
+		delete(data, Region)
+	} else {
+		return fmt.Errorf(RegionMissingError)
+	}
+
+	if secretName := data[SecretName]; secretName != "" {
+		a.secretName = secretName
+		delete(data, SecretName)
+	} else {
+		return fmt.Errorf(SecretNameMissingError)
+	}
+
+	if secretKey := data[SecretKey]; secretKey != "" {
+		a.secretKey = secretKey
+		delete(data, SecretKey)
+	}
+
+	if a.isFile && a.secretKey != "" {
+		return fmt.Errorf(EncryptedFilesShouldNotSpecifyKeyError)
+	}
+
+	if len(data) > 0 {
+		return fmt.Errorf(GenericMalformedKeyError)
+	}
+
+	return nil
+}
+
+func parseBinaryFile(secretValue *secretsmanager.GetSecretValueOutput) (string, error) {
+	return ToTempFile(secretValue.SecretBinary)
+}
+
+func parsePlaintextFile(secretValue *secretsmanager.GetSecretValueOutput) (string, error) {
+	return ToTempFile([]byte(*secretValue.SecretString))
+}
+
+func parseSecretValue(secretValue *secretsmanager.GetSecretValueOutput) (string, error) {
+	return *secretValue.SecretString, nil
+}
+
+func parseSecretKVPair(secretValue *secretsmanager.GetSecretValueOutput, key *string) (string, error) {
+	if secretValue.SecretString == nil {
+		return "", fmt.Errorf(MalformedKVPairSecretPayload)
+	}
+	valueAsByteArray := []byte(*secretValue.SecretString)
+	kvPairs := make(map[string]interface{})
+	err := json.Unmarshal(valueAsByteArray, &kvPairs)
+
+	if err != nil {
+		return "", fmt.Errorf(MalformedKVPairSecretPayload)
+	}
+
+	untypedValue := kvPairs[*key]
+	valueForKeyAsString, ok := untypedValue.(string)
```

Comment:
What is the `.(string)` syntax doing?

Reply:
Type casting / checking in go

if ok == true that means it was a string and that valueForKeyAsString is that value as a string.
if ok == false that means it wasn't a string

Reply:
This is the equivalent of 

```java
Map<String, Object> data = .....
Object thisShouldBeAString = data.get("someKey");
String value
boolean ok = false;
if (thisShouldBeAString instancof String) {
  ok = true;
 value = thisShouldBeAString;
}
if (!ok) {
    // user input was not a string, react
}
```